crates/federate/src/worker.rs

```diff
@@ -137,148 +215,245 @@ impl InstanceWorker {
         return Ok(());
       }
       let remaining = required - elapsed;
+      tracing::debug!(
+        "{}: fail-sleeping for {:?} before starting queue",
+        self.instance.domain,
+        remaining
+      );
       tokio::select! {
         () = sleep(remaining) => {},
         () = self.stop.cancelled() => {}
       }
     }
     Ok(())
   }
-  /// send out a batch of CHECK_SAVE_STATE_EVERY_IT activities
-  async fn loop_batch(&mut self, pool: &mut DbPool<'_>) -> Result<()> {
-    let latest_id = get_latest_activity_id(pool).await?;
-    let mut id = if let Some(id) = self.state.last_successful_id {
-      id
-    } else {
+
+  /// get newest activity id and set it as last_successful_id if it's the first time this instance is seen
+  async fn get_latest_id(&mut self) -> Result<ActivityId> {
+    let latest_id = get_latest_activity_id(&mut self.pool()).await?;
+    if let None = self.state.last_successful_id {
       // this is the initial creation (instance first seen) of the federation queue for this instance
       // skip all past activities:
       self.state.last_successful_id = Some(latest_id);
       // save here to ensure it's not read as 0 again later if no activities have happened
-      self.save_and_send_state(pool).await?;
-      latest_id
-    };
-    if id >= latest_id {
-      // no more work to be done, wait before rechecking
-      tokio::select! {
-        () = sleep(*WORK_FINISHED_RECHECK_DELAY) => {},
-        () = self.stop.cancelled() => {}
+      self.save_and_send_state().await?;
+    }
+    Ok(latest_id)
+  }
+
+  async fn handle_send_results(
+    &mut self,
+    receive_inbox_result: &mut mpsc::UnboundedReceiver<SendActivityResult>,
+    successfuls: &mut BinaryHeap<SendSuccessInfo>,
+    in_flight: &mut i64,
+  ) -> Result<(), anyhow::Error> {
+    let force_write = false;
```

Comment:
unused

Reply:
thanks! It should be set on failure so when a request fails that state is always written to db. Fixed